hoot-core/src/main/cpp/hoot/core/scoring/multiary/MultiaryMatchComparator.cpp

```diff
@@ -0,0 +1,592 @@
+
+      if (intersection.size() > 0)
+      {
+        idsToAdd.append(id);
+        idsToAdd.append(*intersection.begin());
+        foundExpectedPair = true;
+        break;
+      }
+    }
+
+    // for now this only when there are two elements in the review. You'll need to do a bit of
+    // noodling if this fails when we expand to buildings or similar that cause 3 or more elements
+    // in a review.
+    assert (rEids.size() == 2 || !foundExpectedPair);
+
+    // if we didn't find what we were looking for...
+    if (!foundExpectedPair)
+    {
+      // arbitrarily pick the first ID from each cluster. There is no need to report a bad review
+      // for every ID in the cluster. Seems kinda silly.
+      foreach (ElementId eid, rEids)
+      {
+        idsToAdd += *_getAllIds(map->getElement(eid)).begin();
+      }
+    }
+
+    // add an actual review between each of the chosen IDs.
+    for (int i = 0; i < idsToAdd.size(); ++i)
+    {
+      for (int j = i + 1; j < idsToAdd.size(); ++j)
+      {
+        _actualReviews.addReview(idsToAdd[i], idsToAdd[j]);
+      }
+    }
+  }
+}
+
+
+void MultiaryMatchComparator::_findExpectedMatches(const ConstOsmMapPtr& in)
+{
+  class TmpVisitor : public ElementVisitor
+  {
+  public:
+    TmpVisitor(IdToEid& idToEid, QHash<QString, IdClusterPtr>& matchGroups) :
+      _idToEid(idToEid),
+      _matchGroups(matchGroups)
+    {
+    }
+
+    virtual void visit(const ConstElementPtr& e)
+    {
+      if (e->getStatus().isInput())
+      {
+        const Tags& t = e->getTags();
+        QString id = MetadataTags::TrainingId();
+
+        if (t.contains(id))
+        {
+          _idToEid[t[id]] = e->getElementId();
+          if (_matchGroups.contains(t[id]) == false)
+          {
+            _matchGroups.insert(t[id], IdClusterPtr(new IdCluster()));
+          }
+          _matchGroups[t[id]]->insert(t[id]);
+        }
+
+        QString match = MetadataTags::TrainingMatch();
+
+        if (t.contains(match))
+        {
+          if (t[match] == "none")
+          {
+            // pass. No matches.
+          }
+          else if (t[match] == "todo")
+          {
+            LOG_WARN("Got an unexpected todo match. " << e->getElementId());
+          }
+          else if (t.contains(id))
+          {
+            if (_matchGroups.contains(t[match]) == false)
+            {
+              _matchGroups.insert(t[match], IdClusterPtr(new IdCluster()));
+            }
+            // copy all the matches in t[id] into t[match]
+            *_matchGroups[t[match]] += *_matchGroups[t[id]];
+
+            // map this t[id] to the same group as t[match]
+            _matchGroups[t[id]] = _matchGroups[t[match]];
+          }
+          else
+          {
+            LOG_WARN("There is a match on an element without an ID tag. " << e->getElementId());
+          }
+        }
+      }
+    }
+
+  private:
+    IdToEid& _idToEid;
+    QHash<QString, IdClusterPtr>& _matchGroups;
+  };
```

Comment:
I don't think so, these visitors are very specific to the match comparator workflow.

Thanks!